pkg/reconciler/pipelinerun/resources/validate_params.go

```diff
@@ -91,40 +89,59 @@ func ValidateObjectParamRequiredKeys(pipelineParameters []v1beta1.ParamSpec, pip
 
 // ValidateParamArrayIndex validate if the array indexing param reference  target is existent
 func ValidateParamArrayIndex(ctx context.Context, p *v1beta1.PipelineSpec, pr *v1beta1.PipelineRun) error {
-	cfg := config.FromContextOrDefaults(ctx)
-	if cfg.FeatureFlags.EnableAPIFields != config.AlphaAPIFields {
-		return nil
-	}
-
 	arrayParams := extractParamIndexes(p.Params, pr.Spec.Params)
 
 	outofBoundParams := sets.String{}
 
 	// collect all the references
 	for i := range p.Tasks {
-		findInvalidParamArrayReferences(p.Tasks[i].Params, arrayParams, &outofBoundParams)
+		err := findInvalidParamArrayReferences(ctx, p.Tasks[i].Params, arrayParams, &outofBoundParams)
```

Comment:
this is very confusing 🙃 sorry to say that 😢 

`findInvalidParamArrayReferences` ranges over `params` and calls `taskrun.FindInvalidParamArrayReference` which in turn checks if the feature flag is set to `alpha`. If it is, then collect the list of out of bound parameters.

while ranging over a list of tasks, some array references are validated using `findInvalidParamArrayReferences` and some are validated using `taskrun.FindInvalidParamArrayReference` based on a list of params or a 
single param. 

We definitely have a room here for some cleanup, not blocking this PR though.

Reply:
Yes that's true. This part of code is a bit complicated. The reason is array indexing is a kind of `StringReplacements`,
And we have a huge number of places which could use it. e.g. in https://github.com/tektoncd/pipeline/blob/99226de6ef1c3f14e38e9b8cbc1fe0e60753a079/pkg/reconciler/taskrun/resources/apply.go#L293.

So for current validation, I traverse all the possible places to check if we are using array indexing, if used then check feature gate. And then check if it is out of bound of the array.